ug: introduce waves page (#803)
foundriesio/docs: source/user-guide/waves/waves.rst

.. _ref-waves-ug:

Waves and Production Targets
============================

> Waves is the FoundriesFactory mechanism to promote and deliver updates to :ref:`production devices <ref-production-registration>`.

Reviewer: It may be interesting to discuss tags (that can be different from development tags) and device group concepts to give an introduction to the customer.

Author: Tags info is covered in the Considerations section, wdyt about that?
I will link any information on device groups in docs here, but if we don't have anything on device groups we should have another User Guide page for that
@kprosise fyi

Author: Action item to create a Device Group User Guide (FFTK-3870).

It allows deciding which updates are promoted to production, with granular control of how the updates are delivered across the field.
It also provides additional security as only Factory Owners and Admins in possession of TUF targets keys are allowed perform these actions.

Suggested Flow
--------------

> When devices are registered to production, they only update to :ref:`ref-production-targets`.

Author: iiuc we treat as CI Targets x Wave Targets x Production Targets
yes, a device can only be registered to production, not promoted from dev to production

Below you can find the standard flow to promote a target to a Production Target using Waves (``fioctl waves``).

.. note::
   This process can only be performed by Factory Owners and Admins owning the :ref:`Offline TUF Targets keys <ref-offline-targets-keys>`.

.. image:: /_static/userguide/waves/waves-flow.png
   :width: 400
   :align: center
   :alt: Waves subcommands suggested flow

Waves Init (``fioctl waves init``)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

>   * Starts a new Wave on the wanted Target.

Reviewer 2: for some reason I don't like "wanted" here, so I'm suggesting to use "selected" instead. Or candidate. But it's only a suggestion, "wanted" is good.

  * No updates delivered at this point.
  * Wave Status: ``Active``.

Waves Rollout (``fioctl waves rollout``)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

>   * Delivers the update to a specific device group.

Reviewer: specific group , devices (uuids ) or fleet percentagem

  * Multiple rollout commands can be issued for a single wave.
  * Different wave rollouts can happen in parallel.

>   * Optional stage, but recommended.

Reviewer 2: I don't get this bullet point. It's an optional stage (what I understand is that, looking to the diagram, this is there, but is optional), but is recommended, but why is it recommended?

I think we can remove this line. This section/chapter is about the suggested flow, so I understand others are possible. So I would only keep this the same level as the others stages. Or note what other suggested stages are NOT optional.

Author: The context here is that there are customers that are not using device groups, so they can init and complete the wave (for example all devices are updated from offline updates). I still think this is relevant information, so maybe we can rework rather than remove this.
It highly depends on the customer business logic.

Author: I added some more information to this line. Hope that clarifies the optional information.

  * Options for granular control are available at this stage, please check command helper and :ref:`ref-production-targets`.
  * Wave Status: ``Active``.

Waves Complete (``fioctl waves complete``)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  * Delivers the update to all production devices following this wave tag.
  * Only after Wave Complete, this Target becomes a Production Target.
  * Wave Status: ``Complete``.

Waves Cancel (``fioctl waves cancel``)
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

  * Cancels an active wave.
  * Devices that updated prior to Waves Cancel stays in the updated version.
  * This Target does not become a Production Target.
  * Wave Status: ``Canceled``.

.. tip::
   Wave Status can be checked with ``fioctl waves list`` or in the ``Waves`` tab of the Factory.

Below is a summary of Waves and Production Targets status at each wave stage:

> .. list-table:: Waves x Production Targets

Reviewer 2: maybe you can add here which stage is optional

* You may create as many waves as you need for your release strategy. But, there may be only one active wave per device tag at a time. If you need to roll out more than one wave to one tag, either complete or cancel a previous wave to that tag.

* The Wave Rollout step can be skipped by initializing and completing the wave. This delivers the update across all fleet at once.

* Tags in :ref:`ref-ci-targets` do not conflict with tags in Production Targets.

  * This means that both CI and Production Targets can have the same tag and not impact the other class of devices. For example, a production device following the ``release`` tag does not update to a CI Target tagged with ``release`` and vice-versa.

* Up to LmP v95, unexpected downgrades can happen during :ref:`TUF keys rotation <ref-offline-keys>` in case there are devices running on Canceled Waves.

  * During the TUF keys rotation, the TUF metadata is refreshed, causing the devices to receive a fresh version of the targets list. As a Canceled Wave does not produce a Production Target, devices running on this Wave will update to the last Complete Wave, causing a downgrade. This is expected in terms of the TUF specification.

>   * Starting with v95, downgrades are forbidden by default.

Reviewer 3: This is important information. This might need to be flagged or highlighted as a warning or some other mechanism to make it clear that key rotation impacts non completed waves pre v95.

  * Targets become Production Targets after the Wave Complete step.

  * You can get the list of Production Targets with:

  .. code-block::

     fioctl targets list --production --by-tag <tag>

* You may create as many waves as you need for your release strategy. But, there may be only one active wave per device tag at a time. If you need to roll out more than one wave to one tag, either complete or cancel a previous wave to that tag.

* The Wave Rollout step can be skipped by initializing and completing the wave. This delivers the update across all fleet at once.

* Tags in :ref:`ref-ci-targets` do not conflict with tags in Production Targets.

  * This means that both CI and Production Targets can have the same tag and not impact the other class of devices. For example, a production device following the ``release`` tag does not update to a CI Target tagged with ``release`` and vice-versa.

* Up to LmP v95, unexpected downgrades can happen during :ref:`TUF keys rotation <ref-offline-keys>` in case there are devices running on Canceled Waves.

>   * During the TUF keys rotation, the TUF metadata is refreshed, causing the devices to receive a fresh version of the targets list. As a Canceled Wave does not produce a Production Target, devices running on this Wave will update to the last Complete Wave, causing a downgrade. This is expected in terms of the TUF specification.

Reviewer 4: A small correction to make it sound crystal clear.

```suggestion
* During the TUF keys rotation, the TUF root metadata is refreshed, causing the devices to receive a fresh version of the targets list. As a Canceled Wave does not produce a Production Target, devices running on this Wave will update to the last Complete Wave, causing a downgrade. This is expected in terms of the TUF specification.
```

FYI. This line also breaks semantic newlines FWIW.

Author: I'm unsure how to use semantic newlines with bullet points, it seems to break styling
maybe @kprosise can help

  * Starting with v95, downgrades are forbidden by default.

>   * To avoid this behavior in previous LmP versions, complete a wave for the impacted devices prior to the rotation.

Reviewer 4: Are these two bullets related? It is not evident from a text flow.
Is this also related to another bullet about v95? (Even less evident).

Anyway, I do agree with Tim that this deserves a separate attention and proper emphasize.
Does it make sense to grab all v95 related nuances into its own thing?

Author: I was expecting to have all of this related to the v95 note, as this is the behavior customers are concerned with
![Screenshot from 2025-03-14 10-58-20](https://github.com/user-attachments/assets/d570fb37-01e5-43ec-a6ca-549d6d77c0b3)

Author: We will add the v95 related nuances to the release notes page, so I though we could have a bigger warning there

Reviewer 4: This screenshot styling actually looks quite good.
If we can group all v95 related notes here similarly to how it was done in the release notes, that would be a reasonable improvement.

* It is not possible to Cancel a Wave after it has been Completed, but its Target can be removed from the Production Targets list by pruning it. To achieve this, Init then Complete a new Wave for this particular Tag with the `--prune <target-number>` parameter:

  .. code-block::

     fioctl waves init -k ~/path/to/keys/targets.only.key.tgz <new-wave> <target-number> <tag> --prune <target-to-prune>
     fioctl waves complete <new-wave>

Production Workflow
-------------------

> The production workflow and release strategy depend on each use case. Here, we present two approaches for simplified usage.

Reviewer 4: ```suggestion
Your production workflow and release strategy depend on your specific use case.
Here, we present two approaches for simple usage.
```

Clues:
- I kind of understand what "each" means, but it looks off... I find "your" more to the point.
- A word "simplified" means that you make something more simple than it was, right? But there is no comparizon in this context, so it's just "simple".

Keeping the Same Tag
~~~~~~~~~~~~~~~~~~~~

This approach benefits use cases where the whole fleet is expected to be in the same software version.

* Production devices follow a unique general tag, for example ``production``, throughout the entire life cycle.

* When a new release is available, a new wave is created for the ``production`` tag.

* Devices update as soon as this new wave is completed or rolled out to their device group.

Creating New Release Tag
~~~~~~~~~~~~~~~~~~~~~~~~

This approach allows granular controls of the fleet and software versions. It is possible to have multiple stable releases for different usage and versions of the product.

* Production devices follow a release tag, for example ``v1.0``.

> * When a new release is available, a new wave is created for the new release tag, for example ``v1.1``.

Reviewer 4: I am not sure if this is really what we'd like to promote, as we (cloud engineering team) kind of hate how it impacts backend performance and UI representation.
I know that sime customers are using this flow... do we want all to follow?

@doanac @mike-scott 

I'd rather promote a flow where a tag designates static distinct device treats inherent from their hardware or use specifics, not a dynamic version of software they run.
